api/src/check_size.rs

```diff
@@ -11,45 +11,37 @@ use poem::{
 /// This middleware confirms that the Content-Length header is set and the
 /// value is within the acceptable range. It only applies to POST requests.
 pub struct PostSizeLimit {
+    inner: E,
     max_size: u64,
 }
 
 impl PostSizeLimit {
-    pub fn new(max_size: u64) -> Self {
-        Self { max_size }
+    pub fn new(inner: E, max_size: u64) -> Self {
+        Self { inner, max_size }
     }
 }
 
 impl<E: Endpoint> Middleware<E> for PostSizeLimit {
-    type Output = PostSizeLimitEndpoint<E>;
+    type Output = Self;
 
     fn transform(&self, ep: E) -> Self::Output {
-        PostSizeLimitEndpoint {
-            inner: ep,
-            max_size: self.max_size,
-        }
+        Self { inner: ep, max_size: self.max_size }
     }
 }
 
-/// Endpoint for PostSizeLimit middleware.
-pub struct PostSizeLimitEndpoint<E> {
-    inner: E,
-    max_size: u64,
-}
-
 #[async_trait::async_trait]
-impl<E: Endpoint> Endpoint for PostSizeLimitEndpoint<E> {
+impl<E: Endpoint> Endpoint for PostSizeLimit {
     type Output = E::Output;
 
     async fn call(&self, req: Request) -> Result<Self::Output> {
-        if req.method() != Method::POST {
+        if !req.method().is_post() {
             return self.inner.call(req).await;
         }
 
-        let content_length = req
-            .headers()
-            .typed_get::<headers::ContentLength>()
-            .ok_or(SizedLimitError::MissingContentLength)?;
+        let content_length = match req.headers().typed_try_get::<headers::ContentLength>() {
+            Some(content_length) => content_length,
+            None => return Err(SizedLimitError::MissingContentLength.into()),
+        };
```

Comment:
Readability is not quite as good here.  Wondering if you did any benchmarking to see what the optimizations are here?

Reply:
Haven't done any benchmarking yet, will work on it. 
About the readability part, I agree with that.